### ISSUE TITLE
Output binary is $name instead of lib$name.so

### ISSUE THREAD

**Issue by Mygod** [mozilla/rust-android-gradle#30]
`gradle.plugin.org.mozilla.rust-android-gradle:plugin:0.8.3`. Android gradle plugin 4.0.0-beta01.

**Comment by thomcc** [comment]
Can you elaborate here? We should be outputting `lib$name.so` currently.

**Comment by Mygod** [comment]
According to my test, it didn't work until we used some patch like https://github.com/shadowsocks/shadowsocks-android/commit/5ea3e3559adfb7b779ba6f3f633a02b3bcf50145.

**Comment by thomcc** [comment]
Can you elaborate on "didn't work"? Sorry, just trying to understand what the issue is.

**Comment by Mygod** [comment]
Yeah as said in title, the output binary is named `$name` instead of `lib$name.so`, so it does not get bundled into apk.

**Comment by Mygod** [comment]
I understand that the code I changed in #31 is supposed to do exactly that but somehow that did not work.

**Comment by Mygod** [comment]
Okay I guess the issue is that we are compiling a *binary* instead of libraries like most Android apps do. I wonder what's the easiest way to do this with this plugin then. Currently we are using a convoluted workaround: https://github.com/shadowsocks/shadowsocks-android/commit/833b35e7b792cf13709e4e5f1f87c591bffcf062

**Comment by ncalexan** [comment]
Should somebody see this, I think we can support this without a workaround right now using [`targetIncludes`](https://github.com/mozilla/rust-android-gradle/blob/master/README.md#targetincludes).  I'll leave this ticket to include a sample project that does exactly this.

**Comment by Mygod** [comment]
I am not sure how this relates to compiling a binary?

**Comment by ncalexan** [comment]
> I am not sure how this relates to compiling a binary?

If I understand correctly, the issue is that `cargo build debug` will produce `binary`, but the plugin won't package it into `jniLibs` (which is where executables must go, last I checked).  I think you can use `targetIncludes` to package it, but I haven't verified.  Hence, let's have a sample project/test showing how to do it and making sure we don't regress the functionality.  Is that clear?

**Comment by Mygod** [comment]
I think last time I checked, the binary is not extracted from apk if the filename does not match `lib$name.so`. I could check again sometime.

**Comment by ncalexan** [comment]
OK, I took a look at this.  The way I'd like to support binaries is as follows:

1.  Make `libname` optional and add `binname` and/or `binnames`.
2. Make `binnames` add `--bin ...` to the command line as appropriate
3. Follow https://developer.android.com/ndk/guides/wrap-script#packaging_wrapsh and put the executable file into the Java resources.  (I could be convinced that renaming to `binname.so` is enough here.)
4. Verify that the APK does get the binary in the correct `libs/$arch/...` location in the APK.
5. Verify that we can actually execute the binary on device and emulator.
